### PR TITLE
mgr/volumes: Add `fs volume rename` command

### PR DESCRIPTION
**Pull request by ajarr**
The `fs volume rename` command renames the volume, i.e.,
orchestrator MDS service, file system, and optionally the data and
metadata pool of the file system.

Fixes: https://tracker.ceph.com/issues/51162
Signed-off-by: Ramana Raja <rraja@redhat.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by ajarr** [comment]
Missing
- Pending ReleaseNotes
- qa tests
- renaming MDS

**Comment by ajarr** [comment]
Need inputs on the following,

- I've added a flag '--rename-pools' that needs to be set to allow renaming metadata and data pools per convention.

- Renaming of pools work only if the volume/file system has one data pool. The convention (implemented in `fs volume create`) sets the name of the metadata pool as cephfs.<volname>.metadata, and the name of the data pool as cephfs.<volname>.data.  

- The command tries to undo the rename of the pools if the file system rename fails.

@batrick What do you think of the above actions are reasonable?

File: src/pybind/mgr/volumes/fs/operations/volume.py [hunk 1]
```diff
@@ -117,6 +117,58 @@ def delete_volume(mgr, volname, metadata_pool, data_pools):
...
+                f"Failed to rename data pool '{data_pool}' of volume "
+                "'{volname}' to '{new_data_pool}'")
+            r, outb, outs = rename_pool(mgr, new_metadata_pool, metadata_pool)
+            if r != 0:
+                log.error(
+                    f"Failed to rename metadata pool '{new_metadata_pool}' of "
+                    "volume '{volname}' to original name '{metadata_pool}'")
+            return r, outb, outs
+
+    r, outb, outs = rename_filesystem(mgr, volname, newvolname)
```
Comment (batrick) [inline src/pybind/mgr/volumes/fs/operations/volume.py:155]: I think this needs to be a stack based approach: file system creation is last when doing `volume create` so file system rename should be first when doing `volume rename`.

File: src/pybind/mgr/volumes/fs/operations/volume.py [hunk 2]
```diff
@@ -117,6 +117,58 @@ def delete_volume(mgr, volname, metadata_pool, data_pools):
...
+        data_pool = data_pools[0]
+
+        r, outb, outs = rename_pool(mgr, metadata_pool, new_metadata_pool)
+        if r != 0:
+            log.error(
+                f"Failed to rename metadata pool '{metadata_pool}' of volume "
+                "'{volname}' to '{new_metadata_pool}'")
+            return r, outb, outs
+
+        r, outb, outs = rename_pool(mgr, data_pool, new_data_pool)
```
Comment (batrick) [inline src/pybind/mgr/volumes/fs/operations/volume.py:143]: Can we use `MgrModule.rename_pool`?

**Review by batrick — CHANGES_REQUESTED** [review]

File: src/pybind/mgr/volumes/fs/operations/volume.py (hunk 1 above)
Comment (ajarr) [reply src/pybind/mgr/volumes/fs/operations/volume.py]: Done

**Comment by ajarr** [comment]
@batrick PTAL

**Comment by batrick** [comment]
jenkins test make check

File: src/pybind/mgr/volumes/fs/operations/volume.py [hunk 3]
```diff
@@ -117,6 +117,92 @@ def delete_volume(mgr, volname, metadata_pool, data_pools):
     result_str = "metadata pool: {0} data pool: {1} removed".format(metadata_pool, str(data_pools))
     return r, result_str, ""
 
+def rename_volume(mgr, volname, newvolname, rename_pools):
+    """
+    rename volume (orch MDS service, file system, pools)
+    """
+    # Check whether orch managed MDS service already renamed
```
Comment (batrick) [inline src/pybind/mgr/volumes/fs/operations/volume.py:124]: why? (put answer in the comment!)

File: src/pybind/mgr/volumes/fs/operations/volume.py [hunk 4]
```diff
@@ -117,6 +117,92 @@ def delete_volume(mgr, volname, metadata_pool, data_pools):
...
+            return -errno.EINVAL, "", str(e)
+        if completion and completion.result:
+            svc = completion.result[0]
+            placement = svc.spec.placement.pretty_str()
+            create_mds(mgr, newvolname, placement)
+            remove_mds_service = True
+
+    r, outb, outs = rename_filesystem(mgr, volname, newvolname)
+    if r != 0:
+        errmsg = f"Failed to rename file system '{volname}' to '{newvolname}'"
```
Comment (batrick) [inline src/pybind/mgr/volumes/fs/operations/volume.py:164]: What if this fails because the file system was already renamed?

**Review by batrick — CHANGES_REQUESTED** [review]

File: src/pybind/mgr/volumes/fs/operations/volume.py (hunk 3 above)
Comment (ajarr) [reply src/pybind/mgr/volumes/fs/operations/volume.py]: Done.

File: src/pybind/mgr/volumes/fs/operations/volume.py (hunk 4 above)
Comment (ajarr) [reply src/pybind/mgr/volumes/fs/operations/volume.py:166]: rename_filesystem() is idempotent. The method calls `fs rename` command, which is idempotent.

**Comment by ajarr** [comment]
@batrick do you think the volume's pools should be renamed by default? With this PR, you'd have to pass --rename-pools flag to rename pools.

File: src/pybind/mgr/volumes/fs/operations/volume.py [hunk 5]
```diff
@@ -117,6 +117,102 @@ def delete_volume(mgr, volname, metadata_pool, data_pools):
     result_str = "metadata pool: {0} data pool: {1} removed".format(metadata_pool, str(data_pools))
     return r, result_str, ""
 
+def rename_volume(mgr, volname, newvolname):
```
Comment (sebastian-philipp) [inline src/pybind/mgr/volumes/fs/operations/volume.py:120]: I'd like this method to be type checked in order to make changes to the orchestrator API feasible. 
```suggestion
def rename_volume(mgr, volname: str, newvolname: str) -> Tuple[int, str, str]:
```

**Review by batrick — CHANGES_REQUESTED** [review]
I think the last piece here is to make sure that volume renaming works with cephadm deployed clusters. We already have some yaml for it in `fs:cephadm`. So I think just copy that and verify that the new volume's MDS names changed?

**Comment by ajarr** [comment]
@batrick can you check if the cephadm tests and the rest of the PR look OK to you? That was the pending piece. 

The cephadm CephFS tests in master are currently failing due to https://tracker.ceph.com/issues/52279

**Comment by batrick** [comment]
jenkins test api

File: src/pybind/mgr/volumes/fs/fs_util.py [hunk 6]
```diff
@@ -51,6 +61,9 @@ def create_mds(mgr, fs_name, placement):
         return -errno.EINVAL, "", str(e)
     return 0, "", ""
 
+def rename_mds(mgr, fs_name, new_fs_name):
```
Comment (batrick) [inline src/pybind/mgr/volumes/fs/fs_util.py:64]: Can this be removed?

**Review by batrick — CHANGES_REQUESTED** [review]
otherwise LGTM! well done!

**Comment by sebastian-philipp** [comment]
> @batrick can you check if the cephadm test [26cde6a](https://github.com/ceph/ceph/commit/26cde6a30163472777d175844f423bd995a9ffc3) and the rest of the PR look OK to you? The cephadm test was the pending piece.
> 
> The cephadm CephFS tests in master are currently failing due to https://tracker.ceph.com/issues/52279

do you have a link to the teuthology job?

File: qa/suites/fs/cephadm/renamevolume/distro/centos_latest.yaml [hunk 7]
```diff
@@ -0,0 +1 @@
+.qa/distros/supported/centos_latest.yaml
```
Comment (sebastian-philipp) [inline qa/suites/fs/cephadm/renamevolume/distro/centos_latest.yaml:1]: this is not going to work. you'll need to use `0-single-container-host.yaml` as in https://github.com/ceph/ceph/tree/master/qa/suites/rbd/iscsi

**Review by sebastian-philipp — DISMISSED** [review]

File: src/pybind/mgr/volumes/fs/fs_util.py (hunk 6 above)
Comment (ajarr) [reply src/pybind/mgr/volumes/fs/fs_util.py]: Fixed! Thanks!

**Comment by ajarr** [comment]
fs/cephadm tests passed
```
CEPH_BRANCH = wip-ajarr-add-volume-rename
teuthology-suite -v --ceph-repo https://github.com/ceph/ceph-ci.git --suite-repo https://github.com/ceph/ceph-ci.git  -c $CEPH_BRANCH -m $MACHINE_NAME -s fs -p 50 -e $CEPH_QA_MAIL --suite-branch $CEPH_BRANCH --filter-all cephadm -N 5
```
http://pulpito.front.sepia.ceph.com/rraja-2022-01-13_21:56:53-fs-wip-ajarr-add-volume-rename-distro-default-smithi/


task/volumes test passed. 1 failed and it looks unrelated
```
CEPH_BRANCH = wip-ajarr-add-volume-rename
teuthology-suite -v --ceph-repo https://github.com/ceph/ceph-ci.git --suite-repo https://github.com/ceph/ceph-ci.git  -c $CEPH_BRANCH -m $MACHINE_NAME -s fs -p 50 -e $CEPH_QA_MAIL --suite-branch $CEPH_BRANCH --filter-all volumes,basic
```
http://pulpito.front.sepia.ceph.com/rraja-2022-01-13_20:01:52-fs-wip-ajarr-add-volume-rename-distro-default-smithi/

**Comment by alfonsomthd** [comment]
@ajarr AFAIK this feature shouldn't have any impact on Dashboard, is it?

**Comment by ajarr** [comment]
> @ajarr AFAIK this feature shouldn't have any impact on Dashboard, is it?

@alfonsomthd I don't think so. Dashboard team may want to expose this command though in a follow up PR as there is no easy way to rename a CephFS volume.

**Comment by ajarr** [comment]
jenkins test dashboard

**Comment by ajarr** [comment]
@alfonsomthd does the ceph dashboard cephadm e2e tests failure lool related to this PR?

**Comment by alfonsomthd** [comment]
jenkins test dashboard cephadm

**Comment by alfonsomthd** [comment]
> @alfonsomthd does the ceph dashboard cephadm e2e tests failure look related to this PR?

I re-triggered the job.

**Comment by alfonsomthd** [comment]
jenkins test dashboard cephadm

**Comment by alfonsomthd** [comment]
@ajarr the error does not seem related.

**Comment by alfonsomthd** [comment]
jenkins test dashboard cephadm

**Review by batrick — APPROVED** [review]

**Comment by ajarr** [comment]
Thanks, @batrick !!

@vshankar please pull this in to your next testing branch. This PR also fixes  fs/cephadm suite incorrect distro issue . Thanks!

**Comment by batrick** [comment]
jenkins test dashboard cephadm

**Comment by vshankar** [comment]
https://tracker.ceph.com/projects/cephfs/wiki/Master#2022-Feb-09

**Review by vshankar — APPROVED** [review]
LGTM

**Pull request merged**